firestore-leaderboard/extension/functions/src/index.ts

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as admin from "firebase-admin";
+import * as functions from "firebase-functions";
+
+import { FieldValue } from "firebase-admin/firestore";
+import { getEventarc } from "firebase-admin/eventarc";
+
+import config from "./config";
+import * as logs from "./logs";
+
+enum ChangeType {
+  CREATE,
+  DELETE,
+  UPDATE,
+}
+
+// Initialize the Firebase Admin SDK
+admin.initializeApp();
+const db = admin.firestore();
+
+logs.init(config);
+
+/** Setup EventArc Channels */
+const eventChannel =
+  process.env.EVENTARC_CHANNEL &&
+  getEventarc().channel(process.env.EVENTARC_CHANNEL, {
+    allowedEventTypes: process.env.EXT_SELECTED_EVENTS,
+  });
+
+const getChangeType = (
+  change: functions.Change<admin.firestore.DocumentSnapshot>
+): ChangeType => {
+  if (!change.after.exists) {
+    return ChangeType.DELETE;
+  }
+  if (!change.before.exists) {
+    return ChangeType.CREATE;
+  }
+  return ChangeType.UPDATE;
+};
+
+const publishEvent = async (leaderboardPath: string) => {
+  if (eventChannel) {
+    await eventChannel.publish({
+      type: `firebase.extensions.firestore-leaderboard.v1.updated`,
+      data: {
+        documentPaths: leaderboardPath,
+      },
+    });
+  }
+};
+
+const addEntryLeaderboardDocument = async (
```

Comment:
This path is pretty similar to update. Wonder if we should just combine them? And create document on demand if not available yet.

Reply:
Done